<?xml version="1.0" ?>
<sdf version="1.8">
  <world name="world_demo">

    <physics name="1ms" type="ignored">
      <max_step_size>.001</max_step_size>
      <real_time_factor>1.0</real_time_factor>
    </physics>

    <light name="sun" type="directional">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>0.8 0.8 0.8 1</diffuse>
      <specular>0.2 0.2 0.2 1</specular>
      <attenuation>
        <range>1000</range>
        <constant>0.9</constant>
        <linear>0.01</linear>
        <quadratic>0.001</quadratic>
      </attenuation>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <include>
      <pose relative_to="world">0 0 -2 0 0 0</pose>
      <uri>model://mbari_wec</uri>

      <plugin filename="ElectroHydraulicPTO" name="buoy_gazebo::ElectroHydraulicPTO">
        <PrismaticJointName>HydraulicRam</PrismaticJointName>
        <PistonArea>1.375</PistonArea>
        <HydMotorDisp>0.30</HydMotorDisp>
        <RotorInertia>1</RotorInertia>
      </plugin>

      <!-- Add Upper/Lower Polytropic Spring plugin -->
      <plugin filename="PolytropicPneumaticSpring" name="buoy_gazebo::PolytropicPneumaticSpring">
        <JointName>HydraulicRam</JointName>
        <chamber>upper_polytropic</chamber>
        <stroke>2.03</stroke>
        <piston_area>0.0127</piston_area>
        <dead_volume>0.0266</dead_volume>
        <T0>283.15</T0>
        <R_specific>0.2968</R_specific>
        <c_p>1.04</c_p>
        <hysteresis>true</hysteresis>
        <n1>1.2371</n1>
        <n2>1.2307</n2>
        <x1>0.9790</x1>
        <x2>1.1912</x2>
        <P1>410291</P1>
        <P2>410266</P2>
      </plugin>

      <!-- <debug_prescribed_velocity>true</debug_prescribed_velocity>-->
      <plugin filename="PolytropicPneumaticSpring" name="buoy_gazebo::PolytropicPneumaticSpring">
        <JointName>HydraulicRam</JointName>
        <chamber>lower_polytropic</chamber>
        <stroke>2.03</stroke>
        <piston_area>0.0115</piston_area>
        <dead_volume>0.0523</dead_volume>
        <T0>283.15</T0>
        <R_specific>0.2968</R_specific>
        <c_p>1.04</c_p>
        <hysteresis>true</hysteresis>
        <n1>1.2327</n1>
        <n2>1.2518</n2>
        <x1>1.1958</x1>
        <x2>1.3935</x2>
        <P1>1212090</P1>
        <P2>1212140</P2>
      </plugin>
      <!-- <debug_prescribed_velocity>true</debug_prescribed_velocity>-->

      <!-- Heave Cone Hydrodynamics -->
      <plugin filename="ignition-gazebo-hydrodynamics-system"
        name="ignition::gazebo::systems::Hydrodynamics">
        <link_name>HeaveCone</link_name>
        <xDotU>0</xDotU>
        <yDotV>0</yDotV>
        <zDotW>3000</zDotW>  <!-- 3000 open, 10000 close -->
        <kDotP>0</kDotP>
        <mDotQ>0</mDotQ>
        <nDotR>0</nDotR>
        <xUU>0</xUU>
        <xU>0</xU>
        <yVV>0</yVV>
        <yV>0</yV>
        <zWW>2700.10625</zWW>  <!-- 2700.10625 open, 3782.25 close -->
        <zW>0</zW>
        <kPP>0</kPP>
        <kP>0</kP>
        <mQQ>0</mQQ>
        <mQ>0</mQ>
        <nRR>0</nRR>
        <nR>0</nR>
      </plugin>
    </include>
  </world>
</sdf>
